OpenTrons/opentrons-api
# refactor(api, robot-server): Refactor ProtocolRunner per protocol type

# Overview

Closes RSS-200.
create an abstract interface for `ProtocolRunner` and implement per `JsonRunner` `PythonAndLegacyRunner` and `LiveRunner`.

Also fixes a bug that was causing live http runs to finish executing the run after the first command (RSS-214).

# Test Plan

- Upload python 2.14, json v6, legacy python protocols and make sure you're able to analyze protocols, run LPC and run protocols. Check that:
    - all runs are created correctly
    - can be started, paused and stopped
    - run results are correct
- (Optional)Test a LiveRunner: Start an empty run and post protocol-intent commands. Make sure all commands are run successfully. Also test pausing and stopping the protocol.

# Changelog

- Refactor ProtocolRunner to be an abstract and implement per `JsonRunner` `PythonAndLegacyRunner` and `LiveRunner`.
- `create_simulating_runner` is now constructed per file type. 
- Change `ProtocolAnalyzer` to create on the fly the simulating runner instead of injecting it. 
- Changed tests that were using an empty run.

# Review requests

- I (Sanniti) have tested the changes on the dev server but it's important to test on a real robot (ot2/ot3 either works) since run flow control actions can't be tested on the dev server.
- We have talked about how having a runner that can allow execution of batch commands could be useful for future features. I agree that it would be useful but the way the engine & run is architected currently, it would need a more detailed change to the architecture (more importantly, careful testing) than is strictly necessary for run split. So I have not attempted to do that in this PR. We will be able to make that change in the future when the feature is on the roadmap.
- Currently we are setting run_started_at (also in the run response) when there is a play action. with this change we will no longer set this value for setup commands/post commands. this is the only real noticeable change -> might want to check with UI folks. they should use createdAt instead for this case. 

# Risk assessment

Medium-high. need to make sure nothing is effected from the refactor for all protocol types and for posting commands/runs with no protocol.
